File: python/birthday_problem.py
# https://rosettacode.org/wiki/Birthday_problem

import random

def calculate_overlap_chances(n: int, iterations: int = 10_000) -> float:
    success = 0

    for i in range(iterations):
        birthdays = set()
        for j in range(n):
            birthdays.add(random.randint(1, 365))
        if len(birthdays) != n: success += 1

    return (success / iterations) * 100


def main():
    print(f'{calculate_overlap_chances(57, 10_000):.4f}%')

if __name__ == '__main__':
    main()

File: nim/birthday_problem.nim
import std/[strformat, math, random, sets]

proc calc_overlap_chance(n: int, iterations: int = 10_000): float =
    var
      success: int = 0
      birthdays = initHashSet[int]()

    for i in 1..iterations:
        birthdays = initHashSet[int]()
        for j in 1..n:
            randomize()
            birthdays.incl(rand(1..365))
        if birthdays.len != n: success += 1

    return (success / iterations) * 100

proc main() =
    randomize()
    var
      overlap_chance: float = calc_overlap_chance(57, 10_000)
    echo &"Overlap chance: {overlap_chance:.4f}%"

main()
